Memory Allocation

MrMinemeet opened the pull request:
Heap is allocated with a const size set in `Heap.hpp`.
The instance provides the following useful methods:
- `Block* alloc(std::string& type)` - Allocates a Block for the passed type. The overhead `sizeof(Block)` is at the time of this PR 8 Bytes. So. e.g., a type that wands 512 bytes actually uses 520 bytes.
- `bool registerType(const std::string& type, TypeDescriptor& descriptor)` - Registers a type descriptor under a certain name. With the TypeDescriptor registered, Blocks from it can be allocated.
- `void dump()` - Prints information about used/free bytes and used/free blocks. See: https://github.com/MrMinemeet/SystemSoftware_MemoryManagement/blob/8899cbc20561c4c90ac96c1ba9b878fd3413d10d/src/Heap.cpp#L181-L193

MrMinemeet committed:
Made project compile

MrMinemeet committed:
Initial structure for Task 01

Co-authored-by: BioGustav | Tobias <BioGustav@users.noreply.github.com>

MrMinemeet committed:
Add .cmake to gitignore

MrMinemeet committed:
Move structure to different files

Co-authored-by: BioGustav | Tobias <BioGustav@users.noreply.github.com>

MrMinemeet committed:
Creation of Block class, and start of free list alloc stuff

Co-authored-by: BioGustav | Tobias <BioGustav@users.noreply.github.com>

BioGustav committed:
inplace allocation Block

MrMinemeet committed:
Removed files I think are not necessary to be on Github

MrMinemeet committed:
Fix memory allocation

Not sure why "sizeof(Block) is not subtracted again in Heap.cpp:57

Co-authored-by: BioGustav | Tobias <BioGustav@users.noreply.github.com>

BioGustav committed:
Start of deallocation

BioGustav committed:
update .gitignore and remove files

MrMinemeet committed:
Small readability improvement of Block::data

MrMinemeet committed:
Give objectSize field proper name

MrMinemeet committed:
Add TODO to increase actual heap size before submission

MrMinemeet committed:
Make use pointer asterisk and things to be on the left

E.g., "void* …" instead of "void *…"

MrMinemeet committed:
Add allocation from TypeDescriptor and smaller adjustments

MrMinemeet committed:
WIP: Move to actual free list

MrMinemeet committed:
Merge Heap::alloc(size) into Heap::alloc(TypeDescr)

MrMinemeet committed:
Messing arround to get "free" block

MrMinemeet committed:
WIP: Use a proper linked list as taught in class

In class a linked list with the "free" blocks is used. Trying to add this here.

MrMinemeet committed:
Add setting of next pointer

MrMinemeet committed:
Add static method to retrieve freeblock size

The size includes the objSize and next-pointer

MrMinemeet committed:
Use function to set objSize in FreeBlock

MrMinemeet committed:
Some cleanup and casting to make clang happy

MrMinemeet committed:
Add some newlines and help to separate output

MrMinemeet committed:
Make difference between test "Block" and implemented block clear

MrMinemeet committed:
Fix incorrect calulation of leftover size

sizeof(Block) was used twice in the calculation, once for 'requestedTotalSize' and once for 'leftOver' (prev. 'newLength')

MrMinemeet committed:
Add some testing with more data on the heap

Also increase heap to 512 byte for testing

MrMinemeet committed:
Fix incorrect block size size usage

Differentiate between normal and used block and cast accordingly to use correct 'totalSize()'

MrMinemeet committed:
Comment out multi-main test for now

MrMinemeet committed:
Fixed issue with size getters in Block-Class

This results in Heap#toString also working now

MrMinemeet committed:
Implement Heap#dump as specified

It is actually part of "2. Garbage Collector", but I would already find it useful here.

MrMinemeet committed:
Add some todo/fixme to Declarations

MrMinemeet committed:
Remove/Comment out Heap#dealloc

MrMinemeet committed:
Reduce block header size to 8 byte

Instead of using the 'used'-flag, it is also as simple as  to check if the TypeDescriptor points to the DataArea. If this is the case, then it is a FreeBlock.

So instead of having a 8byte pointer + 1byte bool + 7byte padding, it is just the 8byte pointer

MrMinemeet committed:
Fix program crash in non-debug mode

Incorrectly placed pre-processor if caused the type registration to not be  executed. Removed them and it works

BioGustav commented on src/Block.cpp:
```diff
@@ -0,0 +1,73 @@
+#include "FreeBlock.hpp"
+
+/*
+ * Takes 16 bytes in memory
+ * 1 byte for the used flag
+ * 8 bytes for the typeDescriptor pointer
+ * 7 bytes padding
```
since we do not use the 'used'-bit anymore

```suggestion
 * Takes 8 bytes in memory for the typeDescriptor pointer
```

BioGustav replied:
Since we do not use the 'used'-bit anymore, we reduced it to 8 bytes
```suggestion
 * Takes 8 bytes in memory for the typeDescriptor pointer
```

BioGustav commented on src/FreeBlock.hpp:
```diff
@@ -0,0 +1,30 @@
+ * The "typeDescriptor" points to the data part of the block.
+ * The data part contains a "objSize" and a pointer to the next free block.
+ */
+class FreeBlock : public Block {
```
I guess that since this exists, the standard 'Block' is always a 'used' one. Would be easier to see, if there were an interface Block and two implementing classes UsedBlock and FreeBlock.

MrMinemeet replied:
see #3

BioGustav commented on src/FreeBlock.cpp:
```diff
@@ -0,0 +1,91 @@
+}
+/*
+ * At second position of data part, there is a pointer to the next free block.
+ */
```
duplicated comment not needed, this is the getter and setter section for NextFreePointer
```suggestion

```

BioGustav commented on src/FreeBlock.cpp:
```diff
@@ -0,0 +1,91 @@
+	*(void**) tmp = nextFree;
+}
+
+FreeBlock* FreeBlock::getNextFree() {
```
```suggestion
/**
 * @return nullptr if this is the last free block
 */
FreeBlock* FreeBlock::getNextFree() {
```

BioGustav commented on src/FreeBlock.cpp:
```diff
@@ -0,0 +1,91 @@
+	// typeDescriptor is a pointer to the data part of the free block
+	this->typeDescriptor = (TypeDescriptor*) dataPosition();
+	// Size of the data part of the whole free block
+	this->setObjSize((int) sizeof(FreeBlock) + requestedSize);
```
```suggestion
	this->setObjSize(actualSize);
```

BioGustav commented on src/FreeBlock.cpp:
```diff
@@ -0,0 +1,91 @@
+#include "Declarations.hpp"
+#include <iostream>
+
+FreeBlock::FreeBlock(int requestedSize) : Block(nullptr) {// Explicitly null here, and set further below. Otherwise, the debug info would overwrite it.
```
may be good to know how many bytes this takes in memory and why (like Block). This should be will be bigger as Block I guess

BioGustav commented on src/Heap.cpp:
```diff
@@ -0,0 +1,295 @@
+	Block* bCur = (Block*) heap_buffer;
+	while (bCur != nullptr && traversedSize < HEAP_SIZE) {
+		int curBlkSize;
+		if (bCur->isFreeBlock()) {
```
again mixed up?
```suggestion
		if (!bCur->isFreeBlock()) {
```

BioGustav commented on src/Heap.cpp:
```diff
@@ -0,0 +1,295 @@
+	Block* bCur = (Block*) heap_buffer;
+	while (bCur != nullptr && traversedSize < HEAP_SIZE) {
+		int curBlkSize;
+		if (bCur->isFreeBlock()) {
```
isn't this the other way round?
```suggestion
		if (!bCur->isFreeBlock()) {
```

MrMinemeet replied:
Jep, I just replaced `used` with `isFreeBlock()` and forgot the negation

MrMinemeet committed:
Adjustments according to suggestions

MrMinemeet committed:
Add UsedBlock and split existing Block

BioGustav reviewed (approved):
Looks good to me!

MrMinemeet commented:
> Looks good to me!

This one isn't done yet: https://github.com/MrMinemeet/SystemSoftware_MemoryManagement/pull/2#discussion_r1443742256.

I'm working on it in a [separate branch](https://github.com/MrMinemeet/SystemSoftware_MemoryManagement/tree/split_block_parts_to_UsedBlock) and want to merge it to this one before merging the current PR.

MrMinemeet committed:
Temporary fix

Somehow the DEBUG flag causes a SegFault in Heap at line 18. The "fb->ToString()" to be precise

MrMinemeet committed:
Move debug fill from Block to UsedBlock

MrMinemeet committed:
Add information to DEBUG flag

MrMinemeet commented:
Feel free to merge after PR #3 is merged into this branch

MrMinemeet committed:
Merge pull request #3 from MrMinemeet/split_block_parts_to_UsedBlock

Split block parts to used block
